### ISSUE TITLE
Cannot create google_cloudfunctions2_function without providing the location attribute

### ISSUE THREAD

**Issue by AlfatahB** [hashicorp/terraform-provider-google#12487]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment or [link](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) the pull request to this issue.

### Terraform Version

Terraform v1.2.8
- provider registry.terraform.io/hashicorp/google v4.34.0

### Affected Resource(s)

* google_cloudfunctions2_function

### Terraform Configuration Files (if applicable)

```tf
resource "google_storage_bucket" "bucket" {
  name     = "tf-gcf-source"  # Every bucket name must be globally unique
  location = "US"
  uniform_bucket_level_access = true
}

resource "google_storage_bucket_object" "object" {
  name   = "function-source.zip"
  bucket = google_storage_bucket.bucket.name
  source = "function-source.zip"  # Add path to the zipped function source code
}

resource "google_cloudfunctions2_function" "function" {
  name = "function-v2"
  description = "a new function"

  build_config {
    runtime = "nodejs16"
    entry_point = "helloHttp"  # Set the entry point 
    source {
      storage_source {
        bucket = google_storage_bucket.bucket.name
        object = google_storage_bucket_object.object.name
      }
    }
  }

  service_config {
    max_instance_count  = 1
    available_memory    = "256M"
    timeout_seconds     = 60
  }
}
```

### Issue Description

While I was creating the datasource for google_cloudfunctions2_function, I found out that we cannot read the cloudfunctions2_function using the read function of resource file without providing the **location** attribute. When I dig down into this, I also got the same error while making the create calls without providing the **location** attribute:
```
Error: Error creating function: googleapi: Error 400: Malformed name: 'projects/projectname/locations//functions/function-v2'
```
It's working fine if I provide the **location** attribute in the configuration file.

I think the replaceVars function in the resource files doesn't deal with **{{location}}** in the string provided and hence if we don't provide **location** explicitly, than we get the error for the malformed url.

As of now, the probable fix should be to keep the **location** attribute required.

### References

* https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloudfunctions2_function
* b/274818025

**Comment by edwardmedia** [comment]
@AlfatahB the location is required for this resource. Do you set it like other common attributes like project, zone etc. in the provider, or environment variables? Once you provide defaults for them, they thus become optional for most of resources/data source. Does this make sense? 

https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/provider_reference#region

**Comment by AlfatahB** [comment]
@edwardmedia I've provided a region in the provider block. Even after providing region in the provider block, I get the same error i.e. Malformed URL error

**Comment by edwardmedia** [comment]
@melinath shouldn't `location` be required for this resource?

**Comment by melinath** [comment]
It seems like location should probably be a required field.

**Comment by melinath** [comment]
Or we should allow a way to specify a default, like we do for project and some other fields.

**Comment by edwardmedia** [comment]
Changing the label from bug to enhancement for triage 
Should make the location to be a required field?

**Comment by toandm** [comment]
Same issue for me. I provided the region in **providers** config. This resource should get location field from that.

**Comment by rileykarson** [comment]
> Same issue for me. I provided the region in providers config. This resource should get location field from that.

Locations and regions are not always the same thing, so we're now avoiding inferring `location` fields based on the provider fields. Unfortunately, while `region` and `zone` come from a common list of GCE regions/zones, `location` is product, specific and often-but-not-always aligns.

**Comment by ScottSuarez** [comment]
Implemented this change for the 5.0 release.
https://github.com/GoogleCloudPlatform/magic-modules/pull/8928
